File: Dev/C++/Mascaret/src/VEHA/Entity/ShapeSpecification.cpp
#include "MascaretApplication.h"
#include "VEHA/Entity/ShapeSpecification.h"
#include "VEHA/Entity/Vector3.h"
#include "VEHA/Entity/Entity.h"
#include "VEHA/Entity/PointSpecification.h"
#include <stdexcept>
#include "VEHA/Kernel/PrimitiveType.h"
#include "Tools/utils.h"

namespace VEHA
{
ShapeSpecification::ShapeSpecification(const string& name) //protected
:VirtualRealitySpecification(MascaretApplication::getInstance()->getModel()->getBasicType("shape")),_name(name)
{
}
ShapeSpecification::ShapeSpecification(const string& name, const string & url, bool movable,bool recursive,const string& shader)
:VirtualRealitySpecification(MascaretApplication::getInstance()->getModel()->getBasicType("shape")),_name(name)
{
	_url=url;
}
ShapeSpecification::ShapeSpecification(const string& name, const vector<double>& distances,const vector<string> & urls, bool movable,bool recursive,const string& shader)
:VirtualRealitySpecification(MascaretApplication::getInstance()->getModel()->getBasicType("shape")),_name(name)
{
	if(urls.size()>0)
	{
		_url=urls[0];
	}
}
ShapeSpecification::~ShapeSpecification()
{
}

const ValueSpecification& ShapeSpecification::operator= (const ValueSpecification& vs)
{
	throw std::runtime_error("can't clone a ShapeSpecification");
	if (vs.getType() != getType())
		throw std::bad_cast();
	const ShapeSpecification & p= dynamic_cast<const ShapeSpecification&>(vs);
	setScale(p.getScale());
	_url=p.getUrl();
	return *this;
}
shared_ptr<ValueSpecification> ShapeSpecification::clone() const
{
	//throw std::runtime_error("can't clone a ShapeSpecification");
	//	shared_ptr<ShapeSpecification> p(new ShapeSpecification());
	//	p->setPosition(getPosition());
	//	p->setRotation(getRotation());
	//	setScale(p.getScale());
	//	setBody(p.getBody());
	//	return p;
	return shared_ptr<ValueSpecification>();
}

}

File: Dev/CS/Mascaret/Mascaret/VEHA/Entity/ShapeSpecification.cs

using System;
using System.Collections.Generic;

namespace Mascaret
{
    public abstract class ShapeSpecification : VirtualRealitySpecification
    {
        public string name;

        private string url;
        public string Url
        {
            get { return url; }
        }

        public ShapeSpecification(string name)
            : base(MascaretApplication.Instance.Model.getBasicType("shape"))
        {
            this.name = name;
        }

        //Default parameters movable=true, recursive =false, shader=""
        public ShapeSpecification(string name, string url, bool movable, bool recursive, string shader)
            : base(MascaretApplication.Instance.Model.getBasicType("shape"))
        {
            this.name = name;
            this.url = url;
        }

        //Default parameters movable=true, recursive =false, shader=""
        public ShapeSpecification(string name, List<double> distances, List<string> urls, bool movable, bool recursive, string shader)
            : base(MascaretApplication.Instance.Model.getBasicType("shape"))
        {
            this.name = name;
            if (urls.Count > 0)
            {
                this.url = urls[0];
            }
        }

        public abstract void setVisibility(bool v);
        public abstract bool getVisibility();

        public abstract void setScale(Vector3 scale);
        public abstract Vector3 getScale();


        public abstract void setEntity(Entity entity);
        public abstract Entity getEntity();

        // --- Shape Modifier Operations ---
        public abstract Color getColor();
        public abstract void setColor(Color color);
        public abstract void setTransparent(Color color);
        public abstract void growUp(float percent);
        public abstract bool restaureColor(Color color);
        public abstract bool restaureShape();

        public abstract double prepareSpeak(string text);
        public abstract bool speak(string text);

        public abstract double playAnimation(string animationName);

        // --- Operations ---
        public override ValueSpecification clone()
        {
            return null;
        }
    }
}

